cli/internal/helm/helm.go

```diff
@@ -28,6 +28,104 @@ As such, the number of exported functions should be kept minimal.
+	"github.com/edgelesssys/constellation/v2/internal/kms/uri"
+	"github.com/edgelesssys/constellation/v2/internal/kubernetes/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/semver"
+	"github.com/edgelesssys/constellation/v2/internal/versions"
+)
+
+const (
+	// AllowDestructive is a named bool to signal that destructive actions have been confirmed by the user.
+	AllowDestructive = true
+	// DenyDestructive is a named bool to signal that destructive actions have not been confirmed by the user yet.
+	DenyDestructive = false
+)
+
+type debugLog interface {
+	Debugf(format string, args ...any)
+	Sync()
+}
+
+// Client is a Helm client to apply charts.
+type Client struct {
+	factory    *actionFactory
+	cliVersion semver.Semver
+	log        debugLog
+}
+
+// NewClient returns a new Helm client.
+func NewClient(kubeConfigPath string, log debugLog) (*Client, error) {
+	kubeClient, err := kubectl.NewFromConfig(kubeConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("initializing kubectl: %w", err)
+	}
+	actionConfig, err := newActionConfig(kubeConfigPath, log)
+	if err != nil {
+		return nil, fmt.Errorf("creating action config: %w", err)
+	}
+	lister := ReleaseVersionClient{actionConfig}
+	cliVersion := constants.BinaryVersion()
+	factory := newActionFactory(kubeClient, lister, actionConfig, cliVersion, log)
+	return &Client{factory, cliVersion, log}, nil
+}
+
+// Options are options for loading charts.
+type Options struct {
+	Conformance      bool
+	HelmWaitMode     WaitMode
+	AllowDestructive bool
+	Force            bool
+}
+
+// ApplyCharts loads the charts and returns the executor to apply them.
+// TODO(elchead): remove validK8sVersion by putting ValidK8sVersion into config.Config, see AB#3374.
+func (h Client) ApplyCharts(conf *config.Config, validK8sversion versions.ValidK8sVersion, idFile clusterid.File, flags Options, tfOutput terraform.ApplyOutput, serviceAccURI string, masterSecret uri.MasterSecret) (Runner, bool, error) {
+	releases, err := h.loadReleases(conf, masterSecret, validK8sversion, idFile, flags, tfOutput, serviceAccURI)
+	if err != nil {
+		return nil, false, fmt.Errorf("loading Helm releases: %w", err)
+	}
+	h.log.Debugf("Loaded Helm releases")
+	actions, includesUpgrades, err := h.factory.GetActions(releases, flags.Force, flags.AllowDestructive)
+	return &ChartApplyExecutor{actions: actions, log: h.log}, includesUpgrades, err
+}
+
+func (h Client) loadReleases(conf *config.Config, secret uri.MasterSecret, validK8sVersion versions.ValidK8sVersion, idFile clusterid.File, flags Options, tfOutput terraform.ApplyOutput, serviceAccURI string) ([]Release, error) {
+	helmLoader := newLoader(conf, idFile, validK8sVersion, h.cliVersion)
+	h.log.Debugf("Created new Helm loader")
+	return helmLoader.loadReleases(flags.Conformance, flags.HelmWaitMode, secret,
+		serviceAccURI, tfOutput)
+}
+
+// Runner runs the Helm actions.
+type Runner interface {
+	Run(ctx context.Context) error
+}
```

Comment:
```suggestion
type Applier interface {
	Apply(ctx context.Context) error
}
```

Together with the above comment, this would make it more clear where the action is executed.